src/datastore/src/filestore.js

```diff
@@ -232,164 +232,239 @@ export default class FileStore extends NetworkStore {
-      body: metadata,
-      client: this.client
-    });
-    request.headers.set('X-Kinvey-Content-Type', metadata.mimeType);
-
-    // If the file metadata contains an _id then
-    // update the file
-    if (metadata._id) {
-      request.method = RequestMethod.PUT;
-      request.url = url.format({
-        protocol: this.client.apiProtocol,
-        host: this.client.apiHost,
-        pathname: `${this.pathname}/${metadata._id}`
-      });
-    }
+    return metadata;
+  }
+
+  /**
+   * @private
+   */
+  formUploadResponse(data, file) {
+    delete data._expiresAt;
+    delete data._requiredHeaders;
+    delete data._uploadURL;
+    data._data = file;
+    return data;
+  }
+
+  /**
+   * Upload a file.
+   *
+   * @param {Blob|string} file  File content
+   * @param {Object} [metadata={}] File metadata
+   * @param {Object} [options={}] Options
+   * @return {Promise<File>} A file entity.
+   */
+  upload(file, metadata = {}, options = {}) {
+    metadata = this.transformMetadata(file, metadata);
+    let kinveySaveData = null
 
     // Execute the request
-    return request.execute()
-      .then(response => response.data)
-      .then((data) => {
-        const uploadUrl = data._uploadURL;
-        const headers = new Headers(data._requiredHeaders);
-        headers.set('content-type', metadata.mimeType);
-
-        // Delete fields from the response
-        delete data._expiresAt;
-        delete data._requiredHeaders;
-        delete data._uploadURL;
-
-        // Execute the status check request
-        const statusCheckRequest = new NetworkRequest({
-          method: RequestMethod.PUT,
-          url: uploadUrl,
-          timeout: options.timeout
-        });
-        statusCheckRequest.headers.addAll(headers.toPlainObject());
-        statusCheckRequest.headers.set('Content-Range', `bytes */${metadata.size}`);
-        return statusCheckRequest.execute(true)
-          .then((statusCheckResponse) => {
-            Log.debug('File upload status check response', statusCheckResponse);
-
-            if (statusCheckResponse.isSuccess() === false) {
-              throw statusCheckResponse.error;
-            }
-
-            if (statusCheckResponse.statusCode !== 308) {
-              return file;
-            }
-
-            // Upload the file
-            options.start = getStartIndex(statusCheckResponse.headers.get('range'), metadata.size);
-            return this.uploadToGCS(uploadUrl, headers, file, metadata, options);
-          })
-          .then((file) => {
-            data._data = file;
-            return data;
-          });
+    return this.saveFileMetadata(options, metadata)
+      .then((response) => {
+        kinveySaveData = response.data;
+        return this.makeStatusCheckRequest(response.data, metadata, options.timeout);
+      })
+      .then((response) => {
+        Log.debug('File upload status check response', response);
+
+        if (response.isSuccess() === false) {
```

Comment:
I haven't changed this logic, but shouldn't we add the backoff and retry here as well, for the status check request?